File: submodules/vdbfusion/src/vdbfusion/vdbfusion/VDBVolume.cpp
// MIT License
//
// Copyright (c) 2022 Ignacio Vizzo, Cyrill Stachniss, University of Bonn
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "VDBVolume.h"

// OpenVDB
#include <openvdb/Types.h>
#include <openvdb/math/DDA.h>
#include <openvdb/math/Ray.h>
#include <openvdb/openvdb.h>

#include <Eigen/Core>
#include <algorithm>
#include <cmath>
#include <functional>
#include <iostream>
#include <memory>
#include <vector>

namespace {

// Compute the signed distance function (SDF) for a voxel
float ComputeSDF(const Eigen::Vector3d& origin,
                 const Eigen::Vector3d& point,
                 const Eigen::Vector3d& voxel_center) {
    const Eigen::Vector3d v_voxel_origin = voxel_center - origin;
    const Eigen::Vector3d v_point_voxel = point - voxel_center;
    const double dist = v_point_voxel.norm();
    const double proj = v_voxel_origin.dot(v_point_voxel);
    const double sign = proj / std::abs(proj);
    return static_cast<float>(sign * dist);
}

// Get the center of a voxel in world coordinates
Eigen::Vector3d GetVoxelCenter(const openvdb::Coord& voxel, const openvdb::math::Transform& xform) {
    const float voxel_size = xform.voxelSize()[0];
    openvdb::math::Vec3d v_wf = xform.indexToWorld(voxel) + voxel_size / 2.0;
    return Eigen::Vector3d(v_wf.x(), v_wf.y(), v_wf.z());
}

}  // namespace

namespace vdbfusion {

VDBVolume::VDBVolume(float voxel_size, float sdf_trunc, bool space_carving /* = false*/, int max_label /* = 10000*/, int max_points /* = 10000000*/, 
                    pybind11::array_t<int> influence_voxel, 
                    pybind11::array_t<float> influence_voxel_last_weight, 
                    pybind11::array_t<float> influence_voxel_new_weight,
                    pybind11::array_t<int> points_2_voxel_index,
                    pybind11::array_t<float> points_2_voxel_tsdf,
                    pybind11::array_t<float> points_2_voxel_weight)
    : voxel_size_(voxel_size), 
      sdf_trunc_(sdf_trunc), 
      space_carving_(space_carving), 
      max_label(max_label),
      max_points(max_points),
      influence_voxel(influence_voxel),
      influence_voxel_last_weight(influence_voxel_last_weight),
      influence_voxel_new_weight(influence_voxel_new_weight),
      points_2_voxel_index(points_2_voxel_index),
      points_2_voxel_tsdf(points_2_voxel_tsdf),
      points_2_voxel_weight(points_2_voxel_weight),
      ray_label(max_label, 0.0f),
      points_label(max_points, 0),
      global_voxel_index(1),  // 0 is reserved for None
      voxel_coord(max_points, std::vector<int>(3, 0))
{
    // Initialize TSDF grid
    tsdf_ = openvdb::FloatGrid::create(sdf_trunc_);
    tsdf_->setName("D(x): signed distance grid");
    tsdf_->setTransform(openvdb::math::Transform::createLinearTransform(voxel_size_));
    tsdf_->setGridClass(openvdb::GRID_LEVEL_SET);

    // Initialize weights grid
    weights_ = openvdb::FloatGrid::create(0.0f);
    weights_->setName("W(x): weights grid");
    weights_->setTransform(openvdb::math::Transform::createLinearTransform(voxel_size_));
    weights_->setGridClass(openvdb::GRID_UNKNOWN);

    // Initialize label grid
    label_ = openvdb::Int32Grid::create(0);
    label_->setName("L(x): label grid");
    label_->setTransform(openvdb::math::Transform::createLinearTransform(voxel_size_));
    label_->setGridClass(openvdb::GRID_UNKNOWN);

    // Initialize voxel index grid
    voxel_index_ = openvdb::Int32Grid::create(0);
    voxel_index_->setName("F(x): voxel index grid");
    voxel_index_->setTransform(openvdb::math::Transform::createLinearTransform(voxel_size_));
    voxel_index_->setGridClass(openvdb::GRID_UNKNOWN);

    // Example: Access and modify pybind11 array
    // auto influence_voxel_mutable = influence_voxel.mutable_unchecked<2>();
    // printf("points_mutable.shape(0): %d\n", influence_voxel_mutable.shape(0));
    // printf("points_mutable.shape(1): %d\n", influence_voxel_mutable.shape(1));
    // for (ssize_t i = 0; i < influence_voxel_mutable.shape(0); i++) {
    //     for (ssize_t j = 0; j < influence_voxel_mutable.shape(1); j++) {
    //         influence_voxel_mutable(i, j) += 1000;
    //     }
    // }
}

// Extract points and their corresponding voxel indices with weight filtering
std::tuple<std::vector<Eigen::Vector3d>, std::vector<int>> VDBVolume::ExtractPointsAndIndices(float min_weight) const {
    std::vector<Eigen::Vector3d> points;
    std::vector<int> indices;

    auto tsdf_acc = tsdf_->getAccessor();
    auto voxel_index_acc = voxel_index_->getAccessor();
    auto weights_acc = weights_->getAccessor();
    const openvdb::math::Transform& xform = tsdf_->transform();

    for (auto iter = tsdf_->beginValueOn(); iter; ++iter) {
        const openvdb::Coord& voxel = iter.getCoord();
        float tsdf_value = tsdf_acc.getValue(voxel);
        float weight = weights_acc.getValue(voxel);

        if (weight > min_weight) {
            Eigen::Vector3d point = GetVoxelCenter(voxel, xform);
            points.push_back(point);

            int voxel_idx = voxel_index_acc.getValue(voxel);
            indices.push_back(voxel_idx);
        }
    }

    return std::make_tuple(points, indices);
}

// Update TSDF and weights for a given voxel
void VDBVolume::UpdateTSDF(const float& sdf,
                           const openvdb::Coord& voxel,
                           const std::function<float(float)>& weighting_function) {
    using AccessorRW = openvdb::tree::ValueAccessorRW<openvdb::FloatTree>;
    if (sdf > -sdf_trunc_) {
        AccessorRW tsdf_acc = AccessorRW(tsdf_->tree());
        AccessorRW weights_acc = AccessorRW(weights_->tree());
        const float tsdf = std::min(sdf_trunc_, sdf);
        const float weight = weighting_function(sdf);
        const float last_weight = weights_acc.getValue(voxel);
        const float last_tsdf = tsdf_acc.getValue(voxel);
        const float new_weight = weight + last_weight;
        const float new_tsdf = (last_tsdf * last_weight + tsdf * weight) / (new_weight);
        tsdf_acc.setValue(voxel, new_tsdf);
        weights_acc.setValue(voxel, new_weight);
    }
}

// Integrate a grid into the TSDF volume
void VDBVolume::Integrate(openvdb::FloatGrid::Ptr grid,
                          const std::function<float(float)>& weighting_function) {
    for (auto iter = grid->cbeginValueOn(); iter.test(); ++iter) {
        const auto& sdf = iter.getValue();
        const auto& voxel = iter.getCoord();
        this->UpdateTSDF(sdf, voxel, weighting_function);
    }
}

// Integrate a point cloud into the TSDF volume
void VDBVolume::Integrate(const std::vector<Eigen::Vector3d>& points,
                          const Eigen::Vector3d& origin,
                          const std::function<float(float)>& weighting_function) {
    if (points.empty()) {
        std::cerr << "PointCloud provided is empty\n";
        return;
    }

    current_points_size = points.size();
    auto influence_voxel_mutable = influence_voxel.mutable_unchecked<2>();
    auto influence_voxel_last_weight_mutable = influence_voxel_last_weight.mutable_unchecked<2>();
    auto influence_voxel_new_weight_mutable = influence_voxel_new_weight.mutable_unchecked<2>();
    auto points_2_voxel_index_mutable = points_2_voxel_index.mutable_unchecked<1>();
    auto points_2_voxel_tsdf_mutable = points_2_voxel_tsdf.mutable_unchecked<1>();
    auto points_2_voxel_weight_mutable = points_2_voxel_weight.mutable_unchecked<1>();

    const openvdb::math::Transform& xform = tsdf_->transform();
    const openvdb::Vec3R eye(origin.x(), origin.y(), origin.z());

    auto tsdf_acc = tsdf_->getUnsafeAccessor();
    auto weights_acc = weights_->getUnsafeAccessor();
    auto voxel_index_acc = voxel_index_->getAccessor();

    std::fill(points_label.begin(), points_label.end(), 0);
    int idx = 0;

    int max_influence_num = 0;
    int average_influence_num = 0;

    std::for_each(points.cbegin(), points.cend(), [&](const auto& point) {
        const Eigen::Vector3d direction = point - origin;
        openvdb::Vec3R dir(direction.x(), direction.y(), direction.z());
        dir.normalize();

        std::fill(ray_label.begin(), ray_label.end(), 0.0f);
        int label_with_max_weight = 0;
        float max_weight = 0.0f;

        const auto depth = static_cast<float>(direction.norm());
        const float t0 = space_carving_ ? 0.0f : depth - sdf_trunc_;
        const float t1 = depth + sdf_trunc_;

        // Traverse voxels along the ray using DDA
        const auto ray = openvdb::math::Ray<float>(eye, dir, t0, t1).worldToIndex(*tsdf_);
        openvdb::math::DDA<decltype(ray)> dda(ray);

        int influence_num = 0;
        float min_sdf = 10000000;

        do {
            const auto voxel = dda.voxel();
            const auto voxel_center = GetVoxelCenter(voxel, xform);
            const auto sdf = ComputeSDF(origin, point, voxel_center);
            if (sdf > -sdf_trunc_) {
                const float tsdf = std::min(sdf_trunc_, sdf);
                const float weight = weighting_function(sdf);
                if (weight < 0.0f) {
                    printf("Negative weight: %f\n", weight);
                }
                const float last_weight = weights_acc.getValue(voxel);
                const float last_tsdf = tsdf_acc.getValue(voxel);
                const float new_weight = weight + last_weight;
                const float new_tsdf = (last_tsdf * last_weight + tsdf * weight) / (new_weight);
                tsdf_acc.setValue(voxel, new_tsdf);
                weights_acc.setValue(voxel, new_weight);

                // Assign voxel index if not set
                int voxel_idx = voxel_index_acc.getValue(voxel);
                if(voxel_idx == 0) {
                    voxel_index_acc.setValue(voxel, global_voxel_index);
                    voxel_idx = global_voxel_index;
                    global_voxel_index++;
                    if (global_voxel_index > 50000000) {
                        printf("voxel_idx out of range\n");
                        exit(0);
                    }
                    voxel_coord[voxel_idx] = {voxel.x(), voxel.y(), voxel.z()};
                }

                // Update influence_voxel arrays
                if(influence_num < 30){
                    influence_voxel_mutable(idx, influence_num) = voxel_idx;
                    influence_voxel_last_weight_mutable(idx, influence_num) = last_weight;
                    influence_voxel_new_weight_mutable(idx, influence_num) = weight;
                }

                // Update points_2_voxel_index with the voxel having minimum SDF
                if (std::abs(sdf) < min_sdf) {
                    min_sdf = std::abs(sdf);
                    points_2_voxel_index_mutable(idx) = voxel_idx;
                    points_2_voxel_tsdf_mutable(idx) = tsdf;
                    points_2_voxel_weight_mutable(idx) = new_weight;
                }

                // Voting for label (if used)
                // ray_label[current_label] += new_weight;
                // if (ray_label[current_label] > max_weight) {
                //     max_weight = ray_label[current_label];
                //     label_with_max_weight = current_label;
                // }
            
                influence_num++;
            }
        } while (dda.step());

        // Update points_label if label voting is used
        // points_label[idx++] = label_with_max_weight;

        // Track influence statistics if needed
        // if(influence_num > max_influence_num) {
        //     max_influence_num = influence_num;
        // }
        // average_influence_num += influence_num;

        idx++;
    });

    // Example: Print integration statistics
    // printf("Integrating %d points\n", idx);
    // printf("max_influence_num: %d\n", max_influence_num);
    // printf("average_influence_num: %d\n", average_influence_num / points.size());
}

// Update label grid values from old_value to new_value
void VDBVolume::UpdateLabelAttribute(int old_value, int new_value) {
    auto label_acc = label_->getAccessor();
    for (auto iter = label_->beginValueOn(); iter; ++iter) {
        if (iter.getValue() == old_value) {
            iter.setValue(new_value);
        }
    }
}

// Prune TSDF grid based on minimum weight
openvdb::FloatGrid::Ptr VDBVolume::Prune(float min_weight) const {
    const auto weights = weights_->tree();
    const auto tsdf = tsdf_->tree();
    const auto background = sdf_trunc_;
    openvdb::FloatGrid::Ptr clean_tsdf = openvdb::FloatGrid::create(sdf_trunc_);
    clean_tsdf->setName("D(x): Pruned signed distance grid");
    clean_tsdf->setTransform(openvdb::math::Transform::createLinearTransform(voxel_size_));
    clean_tsdf->setGridClass(openvdb::GRID_LEVEL_SET);
    clean_tsdf->tree().combine2Extended(tsdf, weights, [=](openvdb::CombineArgs<float>& args) {
        if (args.aIsActive() && args.b() > min_weight) {
            args.setResult(args.a());
            args.setResultIsActive(true);
        } else {
            args.setResult(background);
            args.setResultIsActive(false);
        }
    });
    return clean_tsdf;
}

// Get the label for each point
const std::vector<int>& VDBVolume::GetPointsLabel() const {
    return points_label;
}

// Get the number of active voxels in the TSDF grid
size_t VDBVolume::GetVoxelCount() const {
    size_t voxel_count = 0;
    for (auto iter = tsdf_->beginValueOn(); iter; ++iter) {
        ++voxel_count;
    }
    return voxel_count;
}

// Get the current global voxel index
size_t VDBVolume::GetGlobalFeatureIdx() const {
    return global_voxel_index;
}

// Example: Access and modify pybind11 array
void VDBVolume::test(pybind11::array_t<float> points) const {
    auto points_mutable = points.mutable_unchecked<2>();
    for (ssize_t i = 0; i < points_mutable.shape(0); i++) {
        for (ssize_t j = 0; j < points_mutable.shape(1); j++) {
            points_mutable(i, j) += 9.0;
        }
    }
}

// Example: Feature vector accessors (commented out)
// std::vector<float> VDBVolume::GetFeature(size_t index) const {
//     if (index >= feature_vector_.size()) {
//         throw std::out_of_range("Index out of range");
//     }
//     return feature_vector_[index];
// }

// void VDBVolume::SetFeature(size_t index, const std::vector<float>& feature) {
//     if (feature.size() != 512) {
//         throw std::invalid_argument("Feature vector must be of size 512");
//     }
//     if (index >= feature_vector_.size()) {
//         throw std::out_of_range("Index out of range");
//     }
//     feature_vector_[index] = feature;
// }

}  // namespace vdbfusion

File: submodules/vdbfusion/src/vdbfusion/pybind/vdb_volume.py
# MIT License
#
# Copyright (c) 2022 Ignacio Vizzo, Cyrill Stachniss, University of Bonn
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import os
from typing import Any, Optional, Tuple, Callable, overload

import numpy as np
import torch
import torch.multiprocessing as mp
import torch.multiprocessing
import open3d as o3d
from . import vdbfusion_pybind


class VDBVolume:
    def __init__(
        self,
        voxel_size: float,
        sdf_trunc: float,
        space_carving: bool = False,
    ):
        # Maximum number of voxels and points
        self.max_points = 3000000
        self.per_max_points = 400000
        self.max_label = 10000

        # Internal buffers for voxel influence and weights
        self.current_points_size = 0
        self.influence_voxel = np.zeros((self.per_max_points, 30), dtype=np.int32)
        self.influence_voxel_last_weight = np.zeros((self.per_max_points, 30), dtype=np.float32)
        self.influence_voxel_new_weight = np.zeros((self.per_max_points, 30), dtype=np.float32)
        self.points_2_voxel_index = np.zeros((self.per_max_points), dtype=np.int32)

        # Global feature matrix for all points
        self.global_clip_feature = torch.zeros((self.per_max_points, 512), dtype=torch.float32, device='cuda').contiguous()
        self.current_points_label = None
        self.current_label_feature = None

        # Initialize C++ backend
        self._volume = vdbfusion_pybind._VDBVolume(
            voxel_size=np.float32(voxel_size),
            sdf_trunc=np.float32(sdf_trunc),
            space_carving=space_carving,
            max_label=self.max_label,
            max_points=self.max_points,
            influence_voxel=self.influence_voxel,
            influence_voxel_last_weight=self.influence_voxel_last_weight,
            influence_voxel_new_weight=self.influence_voxel_new_weight,
            points_2_voxel_index=self.points_2_voxel_index,
        )

        # Passthrough all data members from the C++ API
        self.voxel_size = self._volume._voxel_size
        self.sdf_trunc = self._volume._sdf_trunc
        self.space_carving = self._volume._space_carving
        self.pyopenvdb_support_enabled = self._volume.PYOPENVDB_SUPPORT_ENABLED
        if self.pyopenvdb_support_enabled:
            self.tsdf = self._volume._tsdf
            self.weights = self._volume._weights

        print("VDBVolume initialized")

    ########## Custom Functions ##########

    def update_global_feature(self) -> None:
        """
        Update the global feature matrix using the influence voxel and weights.
        This step distills the influence_voxel array and updates the global_clip_feature tensor.
        """
        mask1 = self.influence_voxel[:self.current_points_size, :] != 0
        _cu_incfluence_voxel = torch.from_numpy(self.influence_voxel[:self.current_points_size, :][mask1]).to(torch.int32).reshape(-1).cuda().contiguous()
        _cu_incfluence_voxel_last_weight = torch.from_numpy(self.influence_voxel_last_weight[:self.current_points_size, :][mask1]).to(torch.float32).reshape(-1).cuda().contiguous()
        _cu_incfluence_voxel_new_weight = torch.from_numpy(self.influence_voxel_new_weight[:self.current_points_size, :][mask1]).to(torch.float32).reshape(-1).cuda().contiguous()
        _cu_current_points_label = self.current_points_label.unsqueeze(1).repeat(1, 30).to(torch.int32)[mask1].reshape(-1).cuda().contiguous()

        # Expand global_clip_feature if needed
        max_voxel_idx = _cu_incfluence_voxel.max()
        if max_voxel_idx >= self.global_clip_feature.size(0):
            self.global_clip_feature = torch.cat(
                (
                    self.global_clip_feature,
                    torch.zeros((max_voxel_idx - self.global_clip_feature.size(0) + 10, 512), dtype=torch.float32, device='cuda').contiguous()
                ),
                dim=0
            )

        # Weighted update of global_clip_feature
        self.global_clip_feature[_cu_incfluence_voxel, :] = (
            self.global_clip_feature[_cu_incfluence_voxel, :] * _cu_incfluence_voxel_last_weight.unsqueeze(1)
            + self.current_label_feature[_cu_current_points_label, :] * _cu_incfluence_voxel_new_weight.unsqueeze(1)
        ) / (
            _cu_incfluence_voxel_last_weight.unsqueeze(1) + _cu_incfluence_voxel_new_weight.unsqueeze(1)
        )

    def set_current_frame_feature(self, points_label, label_feature) -> None:
        """
        Set the current frame's point labels and label features.
        points_label: N x 1 tensor, each point's label
        label_feature: M x 512 tensor, each label's feature
        """
        self.current_points_label = points_label
        self.current_label_feature = label_feature

    ########## Predefined Functions ##########

    def __repr__(self) -> str:
        return (
            f"VDBVolume with:\n"
            f"voxel_size    = {self.voxel_size}\n"
            f"sdf_trunc     = {self.sdf_trunc}\n"
            f"space_carving = {self.space_carving}\n"
        )

    @overload
    def integrate(
        self,
        points: np.ndarray,
        extrinsic: np.ndarray,
        weighting_function: Callable[[float], float],
    ) -> None:
        ...

    @overload
    def integrate(self, points: np.ndarray, extrinsic: np.ndarray, weight: float) -> None:
        ...

    @overload
    def integrate(self, points: np.ndarray, extrinsic: np.ndarray) -> None:
        ...

    @overload
    def integrate(self, grid, weighting_function: Callable[[float], float]) -> None:
        ...

    @overload
    def integrate(self, grid, weight: float) -> None:
        ...

    @overload
    def integrate(self, grid) -> None:
        ...

    def integrate(
        self,
        points: Optional[np.ndarray] = None,
        extrinsic: Optional[np.ndarray] = None,
        grid: Optional[Any] = None,
        weight: Optional[float] = None,
        weighting_function: Optional[Callable[[float], float]] = None,
    ) -> None:
        """
        Integrate new points or grid into the TSDF volume.
        This function supports multiple overloads for different input types.
        """
        self.current_points_size = points.shape[0]
        if self.current_points_size > self.per_max_points or self.current_points_size == 0:
            print(f"Current points size is {self.current_points_size}")
            raise ValueError(f"points size must be in [1, {self.per_max_points}]")

        # Reset influence voxel buffers for current points
        self.influence_voxel[:points.shape[0], :] = 0
        self.influence_voxel_last_weight[:points.shape[0], :] = 0
        self.influence_voxel_new_weight[:points.shape[0], :] = 0

        if grid is not None:
            if not self.pyopenvdb_support_enabled:
                raise NotImplementedError("Please compile with PYOPENVDB_SUPPORT_ENABLED")
            if weighting_function is not None:
                return self._volume._integrate(grid, weighting_function)
            if weight is not None:
                return self._volume._integrate(grid, weight)
            return self._volume._integrate(grid)
        else:
            assert isinstance(points, np.ndarray), "points must be np.ndarray(n, 3)"
            assert points.dtype == np.float64, "points dtype must be np.float64"
            assert isinstance(extrinsic, np.ndarray), "origin/extrinsic must be np.ndarray"
            assert extrinsic.dtype == np.float64, "origin/extrinsic dtype must be np.float64"
            assert extrinsic.shape in [
                (3,),
                (3, 1),
                (4, 4),
            ], "origin/extrinsic must be a (3,) array or a (4,4) matrix"

            _points = vdbfusion_pybind._VectorEigen3d(points)
            if weighting_function is not None:
                return self._volume._integrate(_points, extrinsic, weighting_function)
            if weight is not None:
                return self._volume._integrate(_points, extrinsic, weight)
            self._volume._integrate(_points, extrinsic)

    @overload
    def update_tsdf(
        self, sdf: float, ijk: np.ndarray, weighting_function: Optional[Callable[[float], float]]
    ) -> None:
        ...

    @overload
    def update_tsdf(self, sdf: float, ijk: np.ndarray) -> None:
        ...

    def update_tsdf(
        self,
        sdf: float,
        ijk: np.ndarray,
        weighting_function: Optional[Callable[[float], float]] = None,
    ) -> None:
        """
        Update the TSDF value for a given voxel index.
        """
        if weighting_function is not None:
            return self._volume._update_tsdf(sdf, ijk, weighting_function)
        return self._volume._update_tsdf(sdf, ijk)

    def extract_triangle_mesh(self, fill_holes: bool = True, min_weight: float = 0.0) -> Tuple:
        """
        Extract the triangle mesh from the TSDF volume.
        Returns vertices and triangles arrays.
        Example usage:
            vertices, triangles = integrator.extract_triangle_mesh()
            mesh = o3d.geometry.TriangleMesh(
                o3d.utility.Vector3dVector(vertices),
                o3d.utility.Vector3iVector(triangles),
            )
        """
        vertices, triangles = self._volume._extract_triangle_mesh(fill_holes, min_weight)
        return np.asarray(vertices), np.asarray(triangles)

    def extract_points_and_indices(self, min_weight: float = 0.0) -> Tuple:
        """
        Extract points and their voxel indices from the TSDF volume.
        Returns points and indices arrays.
        """
        points, index = self._volume._extract_points_and_indices(min_weight)
        return np.asarray(points), np.asarray(index)

    def extract_vdb_grids(self, out_file: str) -> None:
        """
        Write the internal map representation to a file.
        Contains both D(x) and W(x) grids.
        """
        self._volume._extract_vdb_grids(out_file)

    def prune(self, min_weight: float):
        """
        Cleanup the TSDF grid using the W(x) weights grid according to a minimum weight threshold.
        This function is ideal to cleanup the TSDF grid before exporting it.
        """
        return self._volume._prune(min_weight)

    def update_label_attribute(self, old_value: int, new_value: int) -> None:
        """
        Update label attributes in the backend.
        """
        self._volume._update_label_attribute(old_value, new_value)

    def get_points_label(self):
        """
        Get the labels of all points in the volume.
        """
        return np.array(self._volume.get_points_label())

    def get_voxel_count(self):
        """
        Get the total number of voxels in the volume.
        """
        return self._volume.get_voxel_count()

    def get_global_feature_idx(self):
        """
        Get the indices of global features.
        """
        return self._volume.get_global_feature_idx()

    def test(self, points: np.ndarray) -> None:
        """
        Test function for backend integration.
        """
        assert points.ndim == 2, "points must be 2D array"
        self._volume._test(points)

    def get_influence_voxel(self) -> torch.Tensor:
        """
        Returns the influence voxel array as a torch tensor.
        """
        return self._volume.get_influence_voxel()

    ########## Save Functions ##########

    def save_points_and_feature(self, out_folder: str) -> None:
        """
        Save the points and feature to files.
        Saves points as .npz and .ply, and features as .npz.
        """
        points, indices = self.extract_points_and_indices(2.0)
        np.savez(os.path.join(out_folder, "points.npz"), points=points)
        pcd = o3d.geometry.PointCloud()
        pcd.points = o3d.utility.Vector3dVector(points)
        o3d.io.write_point_cloud(os.path.join(out_folder, "points.ply"), pcd)

        feature = self.global_clip_feature[indices.astype(np.int32)].cpu().numpy()
        np.savez(os.path.join(out_folder, "feature.npz"), feature=feature)